x/incentives/keeper/hooks_test.go

```diff
@@ -0,0 +1,292 @@
+
+	// System under test
+	err = s.App.IncentivesKeeper.AfterEpochEnd(s.Ctx, distrEpochIdentifier, 4)
+	s.Require().NoError(err)
+
+	// Validate distribution - expected distributions stay the same relative to previous epoch
+
+	// Note that the perpetual gauge was not refunded. As a result, it is not distributing anymore.
+	s.validateDistributionForGroup(perpetualGroupPoolIDs, perpetualPoolIDToExpectedDistributionMap)
+	// Note that this was the last distribution for non-perpetual gauges. As a result they do not distribute anymore.
+	s.validateDistributionForGroup(nonPerpetualGroupPoolIDs, nonPerpetualPoolIDToExpectedDistributionMap)
+
+	// Validate that perpetual gauge is still present
+	s.validateGroupExists(perpetualGroupGaugeID)
+}
+
+// TODO: create the following tests:
+// https://github.com/osmosis-labs/osmosis/issues/6559
+//
+// Test_AfterEpochEnd_Group_OverlappingPoolsInGroups
+// Test_AfterEpochEnd_Group_NoVolumeOnePool_SkipSilent
+// Test_AfterEpochEnd_Group_ChangeVolumeBetween
+// Test_AfterEpochEnd_Group_CreateGroupsBetween
+// Test_AfterEpochEnd_Group_SwapAndDistribute
+
+// for each pool ID, retrieves its internal gauge and asserts that the gauge has coins according to the
+// poolIDToExpectedDistributionMapOne.
+func (s *KeeperTestSuite) validateDistributionForGroup(groupPoolIDs []uint64, poolIDToExpectedDistributionMapOne map[uint64]sdk.Coins) {
+	s.Require().NotEmpty(poolIDToExpectedDistributionMapOne)
+
+	for i, poolID := range groupPoolIDs {
+		gaugeID, err := s.App.PoolIncentivesKeeper.GetInternalGaugeIDForPool(s.Ctx, poolID)
+		s.Require().NoError(err)
+
+		gauge, err := s.App.IncentivesKeeper.GetGaugeByID(s.Ctx, gaugeID)
+		s.Require().NoError(err)
+
+		fmt.Printf("poolID %d gauge %d %s\n", poolID, gaugeID, gauge.Coins)
+
+		// Note that to avoid leaving dust in the gauge, we distribute
+		// all remaning coins to the last gauge.
+		// As a result, we allow error tolerance of 1.
+		if i == len(groupPoolIDs)-1 {
+			// 10 because it accumulates for multi-epoch tests.
+			tolerance := osmomath.ErrTolerance{AdditiveTolerance: osmomath.NewDec(10), RoundingDir: osmomath.RoundUp}
+			osmoassert.Equal(s.T(), tolerance, poolIDToExpectedDistributionMapOne[poolID], gauge.Coins)
+			break
+		}
+
+		s.Require().Equal(poolIDToExpectedDistributionMapOne[poolID].String(), gauge.Coins.String())
+	}
+}
+
+// computes the expected distribution values for each pool in the map based on the volume each one has and the total volume.
+// The expected distribution is calculated pro-rata based on the volume of each pool.
+func (*KeeperTestSuite) computeExpectedDistributonAmountsFromVolume(poolIDToVolumeMap map[uint64]math.Int, totalVolume math.LegacyDec) map[uint64]sdk.Coins {
+	poolIDToExpectedDistributionMapOne := map[uint64]sdk.Coins{}
+	for poolID, volume := range poolIDToVolumeMap {
+		currentDistribution := coins.MulDec(defaultCoins, volume.ToLegacyDec().Quo(totalVolume))
```

Comment:
do you think it should be called something different than coins, maybe coinsutil or something? I embarrassingly was looking around for where the coins object was and finally realized it was just an import lol.

Reply:
Fair point, I think it's outside of the scope of this PR. Will make an issue

Reply:
https://github.com/osmosis-labs/osmosis/issues/6583